### PR TITLE
[SW-1687] Expose Offset Column in Supervised Algorithms

### PR DESCRIPTION
**Pull request by mn-mikke**
This PR depends on https://github.com/h2oai/h2o-3/pull/4086.

### CODE REVIEW

**Comment by jakubhava** [comment]
@mn-mikke as we were discussing with @michalkurka, exposing offsetColumn is one thing, but we also need to modify our mojo scoring, but at this stage I do not understand the steps completely.

But let's do that as part of different PR

**Comment by jakubhava** [comment]
Created JIRA for that ^ https://0xdata.atlassian.net/browse/SW-1707

File: scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala [hunk 1]
```diff
@@ -56,6 +58,22 @@ class H2OMOJOModel(override val uid: String) extends H2OMOJOModelBase[H2OMOJOMod
...
+      flatDataFrame: DataFrame,
+      udfConstructor: Array[String] => UserDefinedFunction,
+      inputs: Array[String]): DataFrame = {
+    val relevantColumnNames = flatDataFrame.columns.intersect(inputs)
+    val args = relevantColumnNames.map(c => flatDataFrame(s"`$c`"))
+    val udf = udfConstructor(relevantColumnNames)
+    val predictWrapper = H2OMOJOCache.getMojoBackend(uid, getMojoData, this)
+    predictWrapper.getModelCategory match {
+      case ModelCategory.Binomial | ModelCategory.Regression | ModelCategory.Multinomial =>
+        flatDataFrame.withColumn(outputColumnName, udf(struct(args: _*), lit(0.0)))
```
Comment (jakubhava) [inline scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala]: I think comment would help for `lit(0.0)`

File: core/build.gradle [hunk 2]
```diff
@@ -30,6 +30,8 @@ dependencies {
 
     // Generic model support
     compile("ai.h2o:h2o-genmodel:${h2oVersion}")
+    // H2O security
+    compile("ai.h2o:h2o-security:${h2oVersion}")
```
Comment (jakubhava) [inline core/build.gradle]: This should be added as part of different PR after upgrade to new H2O :)

Comment (mn-mikke) [reply core/build.gradle]: Yup, no problem. I just wanted to make it running against the current version of the `rel-yau` branch.

File: scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala (hunk 1 above)
Comment (mn-mikke) [reply scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala]: Ok, will add something.

File: core/build.gradle (hunk 2 above)
Comment (jakubhava) [reply core/build.gradle]: Yup

File: scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala (hunk 1 above)
Comment (jakubhava) [reply scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala]: thank you!

File: core/build.gradle (hunk 2 above)
Comment (michalkurka) [reply core/build.gradle]: @mn-mikke since this is an issue for you right now - should there be a dependency in h2o-core on h2o-security even though it doesn't actually need that

I want to avoid issues not just for SW but for other people building H2O as well

Comment (jakubhava) [reply core/build.gradle]: @michalkurka was thinking about this, on Sparkling Water side we simply fix it, but as you mentioned, it can affect other folks. I think it's actually a good idea to mark it as a dependency for maven purposes

Comment (jakubhava) [reply core/build.gradle]: But it's also a little bit hack to create such artificial dependency, not sure

Comment (mn-mikke) [reply core/build.gradle]: IMHO, there are very few customers like us referencing individual h2o projects. The vast majority will assembly jar. Maybe, it would be safer to such a change in a major release.

File: ml/src/test/scala/org/apache/spark/ml/spark/models/H2OSupervisedMOJOModelTestSuite.scala [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+  test("Offset column gets propagated to MOJO model - GBM") {
+    testOffsetColumnGetsPropagatedToMOJOModel(new H2OGBM())
+  }
+
+  test("Offset column gets propagated to MOJO model - GLM") {
+    testOffsetColumnGetsPropagatedToMOJOModel(new H2OGLM())
+  }
+
+  test("Offset column gets propagated to MOJO model - DeepLearning") {
+    testOffsetColumnGetsPropagatedToMOJOModel(new H2ODeepLearning())
```
Comment (jakubhava) [inline ml/src/test/scala/org/apache/spark/ml/spark/models/H2OSupervisedMOJOModelTestSuite.scala]: Is Model offset available also on XGBoost?

Comment (mn-mikke) [reply ml/src/test/scala/org/apache/spark/ml/spark/models/H2OSupervisedMOJOModelTestSuite.scala]: It needs to be implemented on the H2O-3 side first. https://0xdata.atlassian.net/browse/PUBDEV-7029
